src/main/java/org/myrobotlab/service/abstracts/AbstractMotorController.java

```diff
@@ -1,17 +1,224 @@
+    double maxY = (mc.getMaxY() == null)? this.maxY:mc.getMaxY();
+    
+    // motor control has highest precedence
+    Double minInput = mc.getMinInput();
+    Double maxInput = mc.getMaxInput();
+    // motor controller has next order of precedence
+    if (minInput == null) {
+      minInput = this.minInput;
+    }
+    if (maxInput == null) {
+      maxInput = this.maxInput;
+    }
+
+    // motor control has highest precedence
+    Double minOutput = mc.getMinOutput();
+    Double maxOutput = mc.getMaxOutput();
+    // motor controller has next order of precedence
+    if (minOutput == null) {
+      minOutput = this.minOutput;
+    }
+    if (maxOutput == null) {
+      maxOutput = this.maxOutput;
+    }
+    
+    if (minInput != null && in < minInput) {
+      log.warn("clipping input {} to {}", in, minInput);
+      in = minInput;
+    }
+    if (maxInput != null && in > maxInput) {
+      log.warn("clipping input {} to {}", in, maxInput);
+      in = maxInput;
+    }
+    
+    double c = minY + ((in - minX) * (maxY - minY)) / (maxX - minX);
+
+    if (minOutput != null && c < minOutput) {
+      log.warn("clipping output {} to {}", c, minOutput);
+      return minOutput;
+    }
+    if (maxOutput != null && c > maxOutput) {
+      log.warn("clipping output {} to {}", c, maxOutput);
+      return maxOutput;
+    }
+    return c;
+  }
+  
+  // FIXME - if kept they should be put in interface
+  public boolean motorDetach(String name) {
+    if (motors.containsKey(name)) {
+      motors.remove(name);
+      return true;
+    }
+    return false;
+  }
+
+  // FIXME - if kept they should be put in interface
+  public void motorMove(String name) {
+    motorMove((MotorControl) Runtime.getService(name));
+  }
+
+  // FIXME - if kept they should be put in interface
+  public void motorMoveTo(String name, Integer position) {
+    error("not implemented");
+  }
+  
+  @Override
+  public void motorReset(MotorControl motor) {
+    // TODO Auto-generated method stub
+
+  }
+  
+  @Override
+  public void motorMoveTo(MotorControl motor) {
+    // TODO Auto-generated method stub
 
+  }
+
+  @Override
+  public void motorStop(MotorControl motor) {
+    motor.move(0);
+  }
+
+  
+  @Override
+  public void detach() {    
+    for (String name : motors.keySet()) {
+      Motor m = (Motor) Runtime.getService(name);
+      if (m != null) {
+        m.detach(this);
+      }
+    }
+  }
+  
+  // only handles motors - subclasses must use super plus
+  // their own isAttached
+  @Override
+  public boolean isAttached(Attachable service) {
+    return motors.containsKey(service.getName());
+  }
+  
+  @Override
+  public Set<String> getAttached() {
+    return motors.keySet();
+  }
+  
+  public void detach(MotorControl device) {
+    motors.remove(device);
+  }
+  
+  /**
+   * Routing Attach - routes ServiceInterface.attach(service) to appropriate
+   * methods for this class
+   */
+  @Override
+  public void attach(Attachable motor) throws Exception {
```

Comment:
This is a vague interface..  seems like we might want to pass the MotorControl interface rather than just blindly allowing anything that's attachable?  Honestly, I'm not sure how we should handle interfaces like this.  The intended behavior of this method could become ambiguous in some scenarios.

Reply:
As a basic building block ... an end user I hope will become accustom to "attaching" services..
This is a "routing" attach - which the implementation or abstract will "route" to the appropriate attach.
The implementation, or abstract can choose to implement more class definition parameters, or interfaces as it evolves.  Imagine how legos have an almost universal interface, even if what you stick together doesn't make sense to someone else - e.g. my lego build looks like a pile of plastic vs a helicopter